From 4ef1d6aa1ff85bb85d2732b9860f8b99b7571ee8 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 14 Jun 2012 10:46:39 -0700
Subject: [PATCH 03/13] ck: mm enable swaptoken only when swap full

Integrating and porting from:
  http://ck.kolivas.org/patches/3.0/3.4/3.4-ck2/patches/

[The swap token is only useful in conditions of swap thrash, and actually
worsens the common case by causing more swapping. Make it only have an effect
when swap is more than half full.

-ck]

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 include/linux/swap.h |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/include/linux/swap.h b/include/linux/swap.h
index b1fd5c7..a712f06 100644
--- a/include/linux/swap.h
+++ b/include/linux/swap.h
@@ -361,9 +361,10 @@ extern void grab_swap_token(struct mm_struct *);
 extern void __put_swap_token(struct mm_struct *);
 extern void disable_swap_token(struct mem_cgroup *memcg);
 
+/* Only allow swap token to have effect if swap is full */
 static inline int has_swap_token(struct mm_struct *mm)
 {
-	return (mm == swap_token_mm);
+	return (mm == swap_token_mm && vm_swap_full());
 }
 
 static inline void put_swap_token(struct mm_struct *mm)
-- 
1.7.5.4

